### PR TITLE
[AUD-1648] RN Scroll To Top

### PR DESCRIPTION
**Pull request by sliptype**
### Description

* Adds custom `useScrollToTop` hook that supports nested tab navigators (unlike the one provided by react-navigation)
* Triggers scroll to top on long tab press and focused tab press
* Does not scroll to top on profile tab page tab press

https://user-images.githubusercontent.com/19916043/157122124-7c97c617-bc55-478e-acab-e483c3d11f7e.mov

### Dragons

Attaching listeners to a few events, they get cleaned up but could have slight performance/memory impact

### How Has This Been Tested?

iOS simulator

### How will this change be monitored?

Not released yet

### CODE REVIEW

File: packages/mobile/src/components/animated-button/AnimatedButtonProvider.tsx [hunk 1]
```diff
@@ -81,25 +83,23 @@ const AnimatedButton = ({
...
-    setIsPlaying,
-    hasMultipleStates,
-    animationRef
-  ])
+  }, [onPress, isActive, setIsPlaying, hasMultipleStates, animationRef])
+
+  const handleLongPress = useCallback(() => {
+    if (onLongPress) {
+      onLongPress()
+      return
```
Comment (dylanjeffers) [inline packages/mobile/src/components/animated-button/AnimatedButtonProvider.tsx:98]: maybe if/else?

File: packages/mobile/src/hooks/useScrollToTop.ts [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+ * see: https://github.com/react-navigation/react-navigation/issues/8586
+ */
+export const useScrollToTop = (
+  scrollToTop: () => void,
+  disableTopTabScroll = false
+) => {
+  const navigation = useNavigation()
+
+  useFocusEffect(
+    useCallback(() => {
```
Comment (dylanjeffers) [inline packages/mobile/src/hooks/useScrollToTop.ts:23]: can we nest hooks like that?

File: packages/mobile/src/screens/explore-screen/tabs/ForYouTab/ForYouTab.tsx [hunk 3]
```diff
@@ -60,12 +63,20 @@ const tiles = [
...
+  useScrollToTop(() => {
+    ref.current?.scrollTo({
+      y: 0,
+      animated: true
+    })
+  })
+
   return (
-    <ScrollView style={styles.tabContainer}>
+    <ScrollView style={styles.tabContainer} ref={ref}>
```
Comment (dylanjeffers) [inline packages/mobile/src/screens/explore-screen/tabs/ForYouTab/ForYouTab.tsx:75]: would it make sense to have special scroll to top flatlist/scrollview components at some point that have useScrollToTop baked in?

File: packages/mobile/src/screens/explore-screen/tabs/ForYouTab/ForYouTab.tsx [hunk 4]
```diff
@@ -60,12 +63,20 @@ const tiles = [
...
       <TabInfo header={messages.infoHeader} text={messages.infoText} />
       <View style={styles.contentContainer}>
         {tiles.map(tile => (
-          <ColorTile style={styles.tile} key={tile.variant} {...tile} />
+          <ColorTile style={styles.tile} key={tile.title} {...tile} />
```
Comment (dylanjeffers) [inline packages/mobile/src/screens/explore-screen/tabs/ForYouTab/ForYouTab.tsx:79]: nice find

**Review by dylanjeffers — APPROVED** [review]
really awesome work

File: packages/mobile/src/hooks/useScrollToTop.ts (hunk 2 above)
Comment (sliptype) [reply packages/mobile/src/hooks/useScrollToTop.ts:23]: Yeah! The execution order isn't changing, this is essentially like assigning the result of `useCallback` to a a variable and passing that to `useFocusEffect` so it's ok

Comment (dylanjeffers) [reply packages/mobile/src/hooks/useScrollToTop.ts:23]: ah word, yeah that makes sense

File: packages/mobile/src/screens/explore-screen/tabs/ForYouTab/ForYouTab.tsx (hunk 3 above)
Comment (sliptype) [reply packages/mobile/src/screens/explore-screen/tabs/ForYouTab/ForYouTab.tsx:75]: Yeah I was going back and forth on this. We would need one for flatlist, section list, and scroll view. And then we don't necessarily want every instance to support scroll to top. So i opted for just using the hook in relevant places. But I could be convinced to make custom components for sure

**Pull request merged**